aleksandrov.igor/S1

MrMels625 opened the pull request:
Fwd, Fake + Ring
9 баллов

![S1](https://github.com/user-attachments/assets/8e34a398-99ae-47ad-9566-ea58e4577e86)

Commit:
Initial commit for S0

Commit:
fix acceptance

Commit:
credetials header

Commit:
out_credentials implementation

Commit:
unit test template

Commit:
credentials test

Commit:
fix test-main

Commit:
fix impl

Commit:
test suite

Commit:
Introduce SFINAE for input iterator validation

Added enable_if_input_iterator template alias to validate if a
given type is an input iterator. This will be used in the list
implementation

Commit:
Add declarations: List, iterators, comparison, swap

- Declared the List class with all member methods and constructors
- Defined List methods with stubs and TODOs
- Internal Node structure is also defined.
- Added list comparison operators, forward declaration, and swap
- Iterators implemented as template class with bool parameter that
  defines is iterator const or not

Commit:
Add boost test module for S1 and tests

- Covered constructors, assignment operator, push/pop/emplace, assign,
  insert
- Added stubs for other tests

Commit:
whotespace fix

Commit:
untis test template

Commit:
message test

Commit:
fix test.main

Commit:
fix impl

Commit:
whitepace fix

Commit:
test fix

Tuc0Sa1amanka committed:
Initial commit for S0

Tuc0Sa1amanka committed:
fixed acceptance

Tuc0Sa1amanka committed:
credentials header

Tuc0Sa1amanka committed:
out_credentials implementation

Tuc0Sa1amanka committed:
unit test template

Tuc0Sa1amanka committed:
credentials test

Tuc0Sa1amanka committed:
fix test main

Tuc0Sa1amanka committed:
credentials fixed

hahaha-polina committed:
Initial commot for S0

hahaha-polina committed:
Initial commit for S0

hahaha-polina committed:
Credentials declaration and implementation

hahaha-polina committed:
Test-main

hahaha-polina committed:
Test-credentials

hahaha-polina committed:
Fix CG(1.5)

hahaha-polina committed:
Fix author

hahaha-polina committed:
Fix content

hahaha-polina committed:
Fix content 2

Commit:
List to ForwardList

Elektrek committed:
Initial commit for S0

Elektrek committed:
Creating credentials header & test-main

Elektrek committed:
Creating credentials-test)

Elektrek committed:
Realizing credentials.cpp

Commit:
Main + input func

Durshl4k committed:
Initial commit for S0

Commit:
+ test S0

Commit:
without suit

Durshl4k committed:
fix acceptance

Durshl4k committed:
credentials header

Durshl4k committed:
out_credentials implementation

Durshl4k committed:
unit test template

Durshl4k committed:
credentials test

Durshl4k committed:
fix test main

Durshl4k committed:
fix impl

shycoshy committed:
tests

shycoshy committed:
fix no.1

Commit:
initial commit

Commit:
first test

Commit:
first name and last name test

Commit:
test done

Commit:
mainS0 test added

shycoshy committed:
fix no.2

shycoshy committed:
fix no.3

shycoshy committed:
fix no.4

Commit:
typo fix

Commit:
typo fix..

Commit:
Update ForwardList implementation

- Replaced default constructor with explicit initialization of begin_ and end_
- Updated constructors to use push_back instead of emplace_back
- Fixed pop_back|front to handle empty list
- Implemented emplace[_back|front], assign, insert
- Added assertions
- Refactoring
- Added size_ field to provide size() with O(1)
- Refactored existing methods to deal with size_ field

Commit:
Refactor ForwardList test suite

- Replaced std::list with kizhin::ForwardList
- Updated method names according CG
- Refactored existed tests according new list interface
- Improved test coverage for edge cases
- Marked incomplete tests with TODOs
- Divided tests with test suits
- Moved testInvariants to test-utils file

Commit:
Implement comparison operators

OlegZholobov committed:
Initial commit for S0

OlegZholobov committed:
fix acceptance

OlegZholobov committed:
implemented out_credentials

OlegZholobov committed:
added test for out_credentials

GuseynovGuseynGG committed:
inittial commit

GuseynovGuseynGG committed:
fix acceptance

GuseynovGuseynGG committed:
credentials header

GuseynovGuseynGG committed:
out_credentials implementation

OlegZholobov committed:
changed test structure

OlegZholobov committed:
added test suite

OlegZholobov committed:
fixed CG

Elektrek committed:
Adding main

GuseynovGuseynGG committed:
ok

Commit:
Refactor ForwardList

- Removed non-const erase overload
- Removed insert(const&) and (&&) via passing by value. The same with pushBack|Front
- Implemented removeIf
- Refactored assign (via construct and swap)
- Defined splice with TODO stubs

Elektrek committed:
Corretcing from tets to test mistake

ZakirovMS committed:
Initial commit for S0

Elektrek committed:
Rly...? Shift key is a joker card or smth??? Correcting - to _

Elektrek committed:
BOOOST to BOOST.

ZakirovMS committed:
Name output

A1exCRE committed:
Initial commit for S0

ZakirovMS committed:
Output personal data to function

ZakirovMS committed:
unit test implementation

A1exCRE committed:
fix acceptance

A1exCRE committed:
credentials header

A1exCRE committed:
out_credentials implementation

ZakirovMS committed:
Fixing the function and implementing the test-main

A1exCRE committed:
test-main

sharksonk committed:
Initial commit for S0

sharksonk committed:
ouy-credentials implemented

Commit:
Initial commit for S0

Commit:
out credentials

sharksonk committed:
add test

Commit:
testsss

Commit:
add to main

mokkenz1 committed:
Initial commit

Commit:
ahahah

NastyKomar committed:
initial commit

mokkenz1 committed:
credentials header

mokkenz1 committed:
outCrementials implemetation

mokkenz1 committed:
Little fix

mokkenz1 committed:
unit test template

mokkenz1 committed:
credentials test

mokkenz1 committed:
fix test-main

aldyn-d committed:
Initial commit S0

aldyn-d committed:
credentials output

aldyn-d committed:
output in main

aldyn-d committed:
tests

aldyn-d committed:
corrections

Robotophile committed:
done

MrMels625 committed:
Initial commit for S1

Commit:
Implement modifiers with tests: unique, reverse, ....

- Introduced testComparisonInvariants helper function in test-utils
- Implemented tests for: remove, reverse, splice, unique
- Implemented comparison operators tests
- Implemented unique, reverse, splice

Commit:
Introduce io-utils, type-utils, list-utils

Commit:
Introduce main

Richmick committed:
initial commit for S1

Commit:
Introduce fake node before the beginning

Added beforeBegin() method to get iterator before begin()
Supported beforeBegin() in tests

Commit:
Implement merge and sort. Refactor spliceAfter

Refactored spliceAfter: strong exception safety via nodes relinking
Implemented merge & sort in ForwardList
Implemented tests for merge & sort
Added new test for spliceAfter with corner case
Cleaned up according CG as well

MrMels625 committed:
List input & output with standard fwd list

MrMels625 committed:
Lab work using standard forward_list

Richmick committed:
declare List

Richmick committed:
implement iterator

A1exMas1ov committed:
initial commit for S1

A1exMas1ov committed:
main input

mipxe committed:
initial commit for S1

mipxe committed:
add input

Richmick committed:
implement constructors & getters

mipxe committed:
add output

mipxe committed:
fix overflow

A1exMas1ov committed:
input and print for fwd list 1.0

A1exMas1ov committed:
add checks

A1exMas1ov committed:
cg

A1exMas1ov committed:
fix overflow part 1

mipxe committed:
finish main

mipxe committed:
fix getSum

mipxe committed:
create node.hpp

mipxe committed:
create const iterator

mipxe committed:
create iterator

MrMels625 committed:
Fixes

mipxe committed:
create list

Richmick committed:
[WIP] add inserters, emplacers, etc

A1exMas1ov committed:
add reverse

A1exMas1ov committed:
add iterator

A1exMas1ov committed:
add FwdList clear, front empty size

mipxe committed:
finish list

mipxe committed:
fix

mipxe committed:
CG 1.4

mipxe committed:
fix iterators

mipxe committed:
fix empty list

A1exMas1ov committed:
add begin, end, clear, pop, push

Richmick committed:
[WIP] modificate splice

mipxe committed:
yraaa fix tupay error

mipxe committed:
create tests for base interface

Richmick committed:
[WIP] SIGABRT

A1exMas1ov committed:
add copy, move and reverse + my list use

A1exMas1ov committed:
add fill

A1exMas1ov committed:
add const-iterator and make const-print

A1exMas1ov committed:
cg

A1exMas1ov committed:
add tests for constructors

A1exMas1ov committed:
cg

Commit:
Move forward-list to common

Richmick committed:
add inserters, emplacers, etc

A1exMas1ov committed:
copyConstructor test + operators == !=

A1exMas1ov committed:
add tests for operators

A1exMas1ov committed:
add test for modifiers

A1exMas1ov committed:
add test for reverse

mipxe committed:
finish test

mipxe committed:
fix list

mipxe committed:
fix test

A1exMas1ov committed:
add remove, removeIf and tests

A1exMas1ov committed:
add sstream in test

A1exMas1ov committed:
cg

MrMels625 committed:
Node, Iterator, List & Tests added

MrMels625 committed:
Trailing whitespaces fix :|

Richmick committed:
implement sort-based methods except sort

Richmick committed:
implement main task

Richmick committed:
add input protection + cbegin(), cend(), etc

A1exMas1ov committed:
splice 1.0

A1exMas1ov committed:
add assign

Richmick committed:
fix hidden ul overflow from stoul

Richmick committed:
add reverse & operator<=>

Commit:
Refactor according CG

- Updated header guards in common (S1 -> common)
- Marked ForwardList, Node and Iterator as final
- Used auto for begin() calls and other common cases
- Refactored loops with range-based for where its possible
- Introduced outputListWithNewLine function to avoid code duplicating
- Refactored lambdas in tests
- Renamed testInvariants to testForwardListInvariants
- Renamed convert to transpose
- Refactor transpose to use std::move and be more efficient
- Removed extra headers
- Added noexcept in ForwardList front,back,reverse
- Minor refactoring

A1exMas1ov committed:
add range and initializer list constructors

A1exMas1ov committed:
add tests for assign

A1exMas1ov committed:
add inputIterator template

Richmick committed:
implement sort, unsafeMerge, erase

Richmick committed:
implement erase & remove

A1exMas1ov committed:
add eraseAfter + tests

A1exMas1ov committed:
add insert

Richmick committed:
finish implementing & add more tests

MrMels625 committed:
const iterator, some methods added

MrMels625 committed:
const iterator added (now for sure)

A1exMas1ov committed:
add all tests

A1exMas1ov committed:
cg

Richmick committed:
add lists equality checker tool

MrMels625 committed:
List methods fixes

Richmick committed:
add end() decrementing & compare_to_value

Commit:
Add algorithm header to list-comparison

Commit:
Refactor: avoid line breaks

Commit:
Refactor iterators

- Renamed template arguments with CamelCase (according CG)
- Refactored SFINAE: used enable_if_t instead of enable_if
- Added assert in iterator methods
- Added noexcept to certain iterator methods and comparison operators
- Removed extra empty lines

Commit:
Add messages in assertions

Commit:
Refactor loops. Introduce algorithms header

- Introduced algorithms header with lexicographicalCompare
- Refactored loops in main
- Refactored braces split
- Refactored ForwardList constructors with insertAfter instead of loops
- Minor refactoring

Volgarenok commented:
> Fwd, Fake + Ring Пока разбираюсь со своим списком и тестами Планирую делать работу на 5+ баллов main пока написан с использованием std::forward_list

Надо точнее. Считайте это частью задания: должна быть поставлена задача и вы к ней идёте. Если захотите потом доделать на бОльший балл, то сделаете это после

Приложите таблицу с выбранными критериями

MrMels625 committed:
More list methods added

Commit:
Improve code structure: split pushBack and replace lambda

- Added pushBack(const T&) and pushBack(T&&) instead of pushBack(T)
- Replaced lambda in main with functor

Commit:
Refactor naming: file names

In common directory:
- forward-list-fwd-declaration.hpp -> forward-list-fwd.hpp
- algorithms.hpp                   -> algorithm-utils.hpp

Commit:
Refactor: improve naming & syntax, add comparator

- Replaced outputListWithNewLine calls with outputList for clarity
- Removed space before range based for loop colon according CG
- Added comparator as parameter in lexicographicalCompare function

Richmick committed:
remove wrong spaces

MrMels625 committed:
List input & output with my list

MrMels625 committed:
List head/tail logic fix

MrMels625 committed:
Copy and move list done

MrMels625 committed:
Trailing whitespaces fix (I like emacs)

MrMels625 committed:
Some methods and operators added

MrMels625 committed:
whitespaces AAAAAAAAAAAAA

MrMels625 committed:
Whitespace fix?

MrMels625 committed:
List assign method fix

mipxe committed:
fixxx

MrMels625 committed:
Some list methods added, interface fixes

MrMels625 committed:
Fixes

MrMels625 committed:
Yet another list methods added

MrMels625 committed:
List methods safety fixes

MrMels625 committed:
Separation main into functions

bugsnotabunny commented on aleksandrov.igor/S1/cIterator.hpp:
```diff
@@ -0,0 +1,80 @@
+  class List;
+
+  template< typename T >
+  class cIterator: public std::iterator< std::forward_iterator_tag, T >
```
ConstIterator

Не очень удачно вы тут сэкономили

bugsnotabunny commented on aleksandrov.igor/S1/cIterator.hpp:
```diff
@@ -0,0 +1,80 @@
+    const T* operator->() const;
+
+    bool operator!=(const cIterator< T >&) const;
+    bool operator==(const cIterator< T >&) const;
```
В методах noexcept потеряли

bugsnotabunny commented on aleksandrov.igor/S1/cIterator.hpp:
```diff
@@ -0,0 +1,80 @@
+    {}
+    cIterator(const cIterator< T >&) = default;
+    ~cIterator() = default;
+    cIterator< T >& operator=(const cIterator< T >&) = default;
```
У вас тут все специальные члены класса дефолтные. Лучше не объявляйте их вообще.

Тут подробнее пояснил:
https://github.com/Volgarenok/spbspu-labs-2025-aads-a/pull/46#discussion_r1979943270

bugsnotabunny commented on aleksandrov.igor/S1/cIterator.hpp:
```diff
@@ -0,0 +1,80 @@
+  template< typename T >
+  cIterator< T >& cIterator< T >::operator++()
+  {
+    node_ = node_->next;
```
assert(node_)

bugsnotabunny replied:
В остальных местах аналогично

bugsnotabunny commented on aleksandrov.igor/S1/iterator.hpp:
```diff
@@ -0,0 +1,80 @@
+    Iterator< T > operator++(int);
+
+    T& operator*();
+    T* operator->();
```
Эти методы константны - внутрянку итератора вы ведь не изменяете

Volgarenok replied:
Для обычных итераторов нужны две пары методов: константные и не константные. Константные возвращают ссылки/указатели на данные как на константные. Подразумевается, что если уж метод константный, то через него данные ни косвенно, ни непосредственно в итераторе поменять нельзя

bugsnotabunny commented on aleksandrov.igor/S1/list-utils.cpp:
```diff
@@ -0,0 +1,88 @@
+
+namespace aleksandrov
+{
+  unsigned long long calcSum(List< unsigned long long >& list)
```
По константной ссылке принимайте

bugsnotabunny commented on aleksandrov.igor/S1/list-utils.cpp:
```diff
@@ -0,0 +1,88 @@
+      else
+      {
+        throw std::overflow_error("There was an overflow error!");
+      }
```
Вынесите сложение с проверкой в отдельную функцию

bugsnotabunny commented on aleksandrov.igor/S1/list-utils.cpp:
```diff
@@ -0,0 +1,88 @@
+    {
+      maxSize = std::max(maxSize, it->second.size());
+    }
+    return maxSize;
```
Вы никак не отличите максимум из списка, в котором лежат одни нули от максимума пустого списка

bugsnotabunny commented on aleksandrov.igor/S1/list-utils.hpp:
```diff
@@ -0,0 +1,35 @@
+  template< class InputIt1, class InputIt2 >
+  bool lexicographicalCompare(InputIt1 first1, InputIt1 last1, InputIt2 first2, InputIt2 last2)
+  {
+    for (; (first1 != last1) && (first2 != last2); ++first1, (void) ++first2)
```
А зачем войдить?

bugsnotabunny commented on aleksandrov.igor/S1/node.hpp:
```diff
@@ -0,0 +1,27 @@
+
+#include <cstdlib>
+
+namespace aleksandrov
```
Унесите Node в неймспейс detail - он не является частью публичного API

bugsnotabunny commented on aleksandrov.igor/S1/node.hpp:
```diff
@@ -0,0 +1,27 @@
+    Node< T >* next;
+
+    Node():
+      data(),
```
Ваш список не должен требовать от данных быть default constructible

bugsnotabunny commented on aleksandrov.igor/S1/main.cpp:
```diff
@@ -0,0 +1,60 @@
+      unsigned long long sum = calcSum(*it);
+      sumList.pushBack(sum);
+    }
+    catch (const std::overflow_error& e)
```
По всей лабе дёргаете операции, которые бросают исключения, а обрабатываете только тут. Почему?

Volgarenok replied:
Только желательно всё-таки как-то так сделать, чтоб не через каждое выражение try/catch был, а как-то чтоб обработчики не дублировались

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+    template< typename U >
+    void pushFront(U&&);
+    template< typename U >
+    void pushBack(U&&);
```
А зачем тут темплейты?

bugsnotabunny replied:
Это типа чтобы по универсальной ссылке принимать? Вы себе этим темплейтом очень уж гибкий интерфейс сделали, что потенциально может привести к неожиданным для пользователя эффект. Сделайте метод для rvalue ссылки и для const. А этот как деталь реализации можете оставить

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+    void swap(List< T >&) noexcept;
+    void clear() noexcept;
+
+    void remove(const T&);
```
noexcept

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+    void assign(std::initializer_list< T >);
+
+    template< typename U >
+    Iterator< T > insertAfter(cIterator< T >, U&&);
```
Тоже непонятно зачем темплейт

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+    bool operator<=(const List< T >&) const noexcept;
+    bool operator>(const List< T >&) const noexcept;
+    bool operator>=(const List< T >&) const noexcept;
+    void reverse() noexcept;
```
Объявите reverse до операторов

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+
+  template< typename T >
+  List< T >::List():
+    fake_(new Node< T >()),
```
Список не должен требовать от T быть default constructible

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+    clear();
+    fake_ = rhs.fake_;
+    tail_ = rhs.tail_;
+    rhs.fake_ = new Node< T >();
```
Избавьтесь от этой аллокации

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+  template< typename T >
+  Iterator< T > List< T >::end() const noexcept
+  {
+    return empty() ? Iterator< T >(tail_) : Iterator< T >(tail_->next);
```
Сделайте просто iterator из nullptr и всё. В другом методе аналогично

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+  }
+
+  template< typename T >
+  const T& List< T >::cfront() const noexcept
```
Просто front

Volgarenok replied:
Так а константные версии верните

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+  }
+
+  template< typename T >
+  const T& List< T >::cback() const noexcept
```
Просто back

Volgarenok replied:
Тоже константную верните

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+  }
+
+  template< typename T >
+  size_t List< T >::size() const noexcept
```
О(N) операция size()??????? Сделайте так, чтобы она была O(1)

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+    size_t size() const noexcept;
+
+    template< typename U >
+    void pushFront(U&&);
```
Не вижу версий с константными указателями

bugsnotabunny commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,499 @@
+  template< typename T >
+  void List< T >::remove(const T& value)
+  {
+    const auto pred = [&value](const T& rhs)
```
Давайте без лямбд пока что. Сделайте функтор

mipxe committed:
fix insert

A1exMas1ov committed:
some fixes + add comparison operators and tests

A1exMas1ov committed:
fixes

A1exMas1ov committed:
cg

A1exMas1ov committed:
fix overflow

MrMels625 committed:
Fixes from remarks 1.0

MrMels625 committed:
My favorite whitespaces...

MrMels625 committed:
Tab fix... ehhhh died from cringe

MrMels625 committed:
Little uninitialized value fix

MrMels625 committed:
Fixes from remarks 1.1

MrMels625 committed:
Little fix

A1exMas1ov committed:
fix

Commit:
Move functor from main to list-utils

Volgarenok commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,603 @@
+#include "list-utils.hpp"
+#include <initializer_list>
+#include <cassert>
+#include <cstdlib>
```
CG

Volgarenok commented on aleksandrov.igor/S1/iterator.hpp:
```diff
@@ -0,0 +1,82 @@
+    Iterator(detail::Node< T >* node):
+      node_(node)
+    {}
+
```
Этот конструктор деталь реализации. Его не должно быть в открытом интерфейсе

Volgarenok commented on aleksandrov.igor/S1/iterator.hpp:
```diff
@@ -0,0 +1,82 @@
+    Iterator():
+      node_(nullptr)
+    {}
+    Iterator(detail::Node< T >* node):
```
explicit

Volgarenok commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,603 @@
+    pos.node_->next = newNode;
+    ++size_;
+    return Iterator< T >(newNode);
+  }
```
Ну так у вас дублируется. Вам ж сказали оставить в деталях реализации метод с универсальной ссылкой. Тут как раз сможете perfect forwarding применить

Volgarenok commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,603 @@
+      fake_->next = newNode;
+    }
+    tail_ = newNode;
+    ++size_;
```
Тоже дублируется

Volgarenok commented on aleksandrov.igor/S1/list.hpp:
```diff
@@ -0,0 +1,603 @@
+  template< typename T >
+  class ConstIterator;
+  template< typename T >
+  class Equalizer;
```
Ну это в `detail` наверное надо

Volgarenok replied:
Не понял зачем вы его тут объявили

Volgarenok commented on aleksandrov.igor/S1/list-utils.hpp:
```diff
@@ -0,0 +1,52 @@
+#define LIST_UTILS_HPP
+
+#include "list.hpp"
+#include <ios>
```
CG!

Volgarenok commented on aleksandrov.igor/S1/list-utils.cpp:
```diff
@@ -0,0 +1,129 @@
+    catch (const std::bad_alloc&)
+    {
+      toTranspose.clear();
+      throw;
```
А зачем вы очищаете? Что это даст? Деструктор у списка потом не отработает или что?

mipxe committed:
fix

mipxe committed:
fix splice and insert

A1exMas1ov committed:
add namespace detail

MrMels625 committed:
Merge remote-tracking branch 'upstream/master'

MrMels625 committed:
Merge branch 'master' into aleksandrov.igor/S1

MrMels625 committed:
Merge branch 'master' into aleksandrov.igor/S1

MrMels625 committed:
9/9 implementation & fixes from remarks 1.2

New list methods added:
- emplaceBack, emplaceFront, emplaceAfter
- unique, unique [pred]
- merge, merge [comp]
- sort, sort [comp]
- private methods spliceAfterImpl, mergeImpl added
Other:
- tests for list completely remade & improved
- splice list method renamed to spliceAfter
- class Equalizer moved from header to removeIf impl
- some list methods impl changed
- bool operators moved to the end of declaration
- useless clear in utils impl removed
- iterators little fixes
- CG 47 fix

MrMels625 committed:
New inserts added & some little fixes

New methods added:
- insert [count]
- insert [range]
- insert [initializer_list]
Other:
- tests for insert and erase improved
- other tests a little bit changed
- minor method signature & impl changes

MrMels625 commented:
### **История ПР навернулась из-за неудачных попыток подтягивания мастера** ☠️

_P.S. доделал на 9 баллов_

Volgarenok commented on aleksandrov.igor/S1/list-utils.cpp:
```diff
@@ -0,0 +1,125 @@
+    catch (const std::bad_alloc&)
+    {
+      throw;
+    }
```
Зачем писать такие обработчики?

Volgarenok commented on aleksandrov.igor/S1/iterator.hpp:
```diff
@@ -0,0 +1,89 @@
+    Iterator< T > operator++(int);
+
+    T& operator*() const noexcept;
+    T* operator->() const noexcept;
```
Ну тогда это наверное не константные методы. Может нужны и те, и другие?

MrMels625 committed:
Fixes from remarks 2.0

- useless try catch blocks removed
- minor iterator's operator * and -> fixes
- push replaced with emplace due to its coolness
- minor visual fixes

MrMels625 committed:
Very important fix

- 4 whitespaces removed

MrMels625 committed:
Merge branch 'master' into aleksandrov.igor/S1
